processor/aggregationprocessor/processor.go

```diff
@@ -0,0 +1,340 @@
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+type aggregationProcessor struct {
+	logger   *zap.Logger
+	mux      *sync.Mutex
+	wg       *sync.WaitGroup
+	doneChan chan struct{}
+	//for mocking in test
+	now func() time.Time
+
+	includeRegex           *regexp.Regexp
+	flushInterval          time.Duration
+	aggregationPeriodStart pcommon.Timestamp
+	aggregationConfs       []AggregateConfig
+	// map resource hash to resourceAggregation
+	aggregationMap map[uint64]*resourceMetadata
+	nextConsumer   consumer.Metrics
+}
+
+func newAggregationProcessor(logger *zap.Logger, cfg *Config, consumer consumer.Metrics) (*aggregationProcessor, error) {
+	regex, err := regexp.Compile(cfg.Include)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile include regex: %w", err)
+	}
+	return &aggregationProcessor{
+		logger:                 logger,
+		mux:                    &sync.Mutex{},
+		wg:                     &sync.WaitGroup{},
+		doneChan:               make(chan struct{}),
+		now:                    time.Now,
+		includeRegex:           regex,
+		flushInterval:          cfg.Interval,
+		aggregationPeriodStart: pcommon.NewTimestampFromTime(time.Now()),
+		aggregationMap:         make(map[uint64]*resourceMetadata),
+		aggregationConfs:       cfg.AggregationConfigs(),
+		nextConsumer:           consumer,
+	}, nil
+}
+
+func (sp *aggregationProcessor) Start(_ context.Context, _ component.Host) error {
+	sp.wg.Add(1)
+	go sp.flushLoop()
+	return nil
+}
+
+func (sp *aggregationProcessor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) error {
+	sp.aggregateMetrics(md)
+	if md.ResourceMetrics().Len() != 0 {
+		// Forward metrics we didn't consume
+		return sp.nextConsumer.ConsumeMetrics(ctx, md)
+	}
+
+	return nil
+}
+
+// Add metrics that we care about to our aggregation.
+// The incoming pmetric.Metrics is modified, such that aggregated metrics are removed.
+func (sp *aggregationProcessor) aggregateMetrics(md pmetric.Metrics) {
+	sp.mux.Lock()
+	defer sp.mux.Unlock()
+
+	rms := md.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		rm := rms.At(i)
+		resAttrs := rm.Resource().Attributes()
+		resKey := mapKey(resAttrs)
+
+		sms := rm.ScopeMetrics()
+		for j := 0; j < sms.Len(); j++ {
+			sm := sms.At(j)
+			ms := sm.Metrics()
+			for k := 0; k < ms.Len(); k++ {
+				m := ms.At(k)
+				if !canAggregateMetric(m) {
+					continue
+				}
+
+				// Metric must match regex
+				if !sp.includeRegex.MatchString(m.Name()) {
+					continue
+				}
+
+				ma := sp.metricAggregation(m, resKey, resAttrs)
+
+				dps := datapointsFromMetric(m)
+				// We remove datapoints that we aggregate here, so we use RemoveIf to iterate the datapoints
+				dps.RemoveIf(func(dp pmetric.NumberDataPoint) bool {
+					switch dp.ValueType() {
+					case pmetric.NumberDataPointValueTypeInt:
+					case pmetric.NumberDataPointValueTypeDouble:
+					default:
+						// ignore empty datapoints
+						return false
+					}
```

Comment:
Does this mean we remove if `dp.ValueType()` is `pmetric.NumberDataPointValueTypeInt` or `pmetric.NumberDataPointValueTypeDouble`? If so an if statement might be clearer. 

```go
if dp.ValueType() != pmetric.NumberDataPointValueTypeDouble && dp.ValueType() != pmetric.NumberDataPointValueTypeInt {
  return false
}

sp.aggregateDatapoint(ma, dp)
return true
```